pkg/chains/formats/slsa/v2alpha2/internal/resolved_dependencies/resolved_dependencies.go

```diff
@@ -0,0 +1,227 @@
+		}
+		resolvedDependencies = append(resolvedDependencies, &rd)
+	}
+
+	// add resolved dependencies from pipeline tasks
+	resolvedDependencies, err = addPipelineTask(logger, pro, resolvedDependencies)
+	if err != nil {
+		return nil, err
+	}
+
+	// add resolved dependencies from pipeline results
+	mats := material.AddMaterialsFromPipelineParamsAndResults(ctx, pro, []common.ProvenanceMaterial{})
+	// convert materials to resolved dependencies
+	resolvedDependencies = append(resolvedDependencies, convertMaterialsToResolvedDependencies(mats, inputResultName)...)
+
+	// remove duplicate resolved dependencies
+	resolvedDependencies, err = removeDuplicateResolvedDependencies(resolvedDependencies)
+	if err != nil {
+		return nil, err
+	}
+	return resolvedDependencies, nil
+}
+
+// convertMaterialToResolvedDependency converts a SLSAv0.2 Material to a resolved dependency
+func convertMaterialsToResolvedDependencies(mats []common.ProvenanceMaterial, name string) []*v1.ResourceDescriptor {
+	rds := []*v1.ResourceDescriptor{}
+	for _, mat := range mats {
+		rd := v1.ResourceDescriptor{}
+		rd.URI = mat.URI
+		rd.Digest = mat.Digest
+		if len(name) > 0 {
+			rd.Name = name
+		}
+		rds = append(rds, &rd)
+	}
+	return rds
+}
+
+// removeDuplicateResolvedDependencies removes duplicate resolved dependencies from the slice of resolved dependencies.
+// Original order of resolved dependencies is retained.
+func removeDuplicateResolvedDependencies(resolvedDependencies []*v1.ResourceDescriptor) ([]*v1.ResourceDescriptor, error) {
+	out := make([]*v1.ResourceDescriptor, 0, len(resolvedDependencies))
+
+	// make map to store seen resolved dependencies
+	seen := map[string]bool{}
+	for _, resolvedDependency := range resolvedDependencies {
+		rDep := v1.ResourceDescriptor{}
+		rDep.URI = resolvedDependency.URI
+		rDep.Digest = resolvedDependency.Digest
+		// This allows us to ignore dependencies that have the same uri and digest.
+		rd, err := json.Marshal(rDep)
+		if err != nil {
+			return nil, err
+		}
+		if seen[string(rd)] {
+			// We dont want to remove the top level pipeline/task config from the resolved dependencies
+			// because its critical to provide that information in the provenance. In SLSAv0.2 spec,
+			// we would put this in invocation.ConfigSource. In order to ensure that it is present in
+			// the resolved dependencies, we dont want to skip it if another resolved dependency from the same
+			// uri+digest pair was already included before.
+			if resolvedDependency.Name == taskConfigName || resolvedDependency.Name == pipelineConfigName {
+				out = append(out, resolvedDependency)
+			} else {
+				continue
+			}
+		}
+		seen[string(rd)] = true
+		out = append(out, resolvedDependency)
+	}
+	return out, nil
+}
+
+// addPipelineTask adds the resolved dependencies from pipeline tasks.
+func addPipelineTask(logger *zap.SugaredLogger, pro *objects.PipelineRunObject, resolvedDependencies []*v1.ResourceDescriptor) ([]*v1.ResourceDescriptor, error) {
```

Comment:
Just mutate the param?
```suggestion
func addPipelineTask(logger *zap.SugaredLogger, pro *objects.PipelineRunObject, resolvedDependencies []*v1.ResourceDescriptor) error {
```